src/Compilers/CSharp/Portable/Utilities/DeclarationDiagnosticBag.cs

```diff
@@ -0,0 +1,46 @@
+﻿using System;
+using System.Diagnostics;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    /// <summary>
+    /// Takes declaration diagnostics and either stores them directly or acts as a proxy to the compilation's
+    /// <see cref="CSharpCompilation.DeclarationDiagnostics"/>, depending on construction.
+    /// </summary>
+    internal struct DeclarationDiagnosticStore
```

Comment:
This seems like a lazy `DiagnosticBag`. Perhaps rename to something more general such as `DiagnosticBagProvider`. Or better yet, does this even need to be lazy? Could the callers simply pass in `Compilation.DeclarationDiagnostics` in the cases where that `DiagnosticBag` should be used?